### ISSUE TITLE
Generic instances for value classes

### ISSUE THREAD

**Issue by pjan** [circe/circe#469]
It would be great if these would make it to the library. Happy to open a PR for it, if somebody points me to the right place for these:

```scala
  implicit def decoderValueClass[T <: AnyVal, V](
    implicit
    gen:     Lazy[Generic.Aux[T, V :: HNil]],
    decoder: Decoder[V]
  ): Decoder[T] = Decoder.instance { cursor ⇒
    decoder(cursor).map { value ⇒
      gen.value.from(value :: HNil)
    }
  }

  implicit def encoderValueClass[T <: AnyVal, V](
    implicit
    gen:     Lazy[Generic.Aux[T, V :: HNil]],
    encoder: Encoder[V]
  ): Encoder[T] = Encoder.instance { value ⇒
    encoder(gen.value.to(value).head)
  }
```

**Comment by travisbrown** [comment]
I'd say let's go either with this (or something using Shapeless's `Unwrapped`) in circe-generic-extras for now.

I'd support making it the default there for fully automatic derivation, and if it works well it could graduate to circe-generic in an upcoming version.

**Comment by cakper** [comment]
👍 for adding it to circe-generic :)

**Comment by kciesielski** [comment]
@pjan I tried to use this but I'm getting compilation error:
```
package.scala:42: wrong number of type arguments for ::, should be 1
[error]       implicit gen: Lazy[Generic.Aux[T, V :: HNil]],
```
in the `implicit gen: Lazy[Generic.Aux[T, V :: HNil]],` line. Circe 0.7.0, shapeless 2.3.2, Scala 2.12.1. I'm not very good with Shapeless, can you give me a hint what can I fix to get this thing working? I managed to find a compilable encoder:
```scala
  implicit def encodeAnyVal[A <: AnyVal, L <: HList, T](
      implicit gen: Generic.Aux[A, L],
      hcons: IsHCons.Aux[L, T, HNil],
      anyval: Encoder[T]
  ): Encoder[A] =
    anyval.contramap(a => hcons.head(gen.to(a)))
```
and I need a decoder.

**Comment by radusw** [comment]
@kciesielski try this:

```scala
  implicit def decodeAnyVal[T, U](
    implicit
    ev: T <:< AnyVal,
    unwrapped: Unwrapped.Aux[T, U],
    decoder: Decoder[U]): Decoder[T] = Decoder.instance[T] { cursor =>

    decoder(cursor).map(value => unwrapped.wrap(value))
  }

  implicit def encodeAnyVal[T, U](
    implicit
    ev: T <:< AnyVal,
    unwrapped: Unwrapped.Aux[T, U],
    encoder: Encoder[U]): Encoder[T] = Encoder.instance[T] { value =>

    encoder(unwrapped.unwrap(value))
  }
```

**Comment by johannegger** [comment]
i'd also really like to see this in circe-generic! is there currently an easy way to have this behaviour AND use fully automatic derivation? (using the anyVal decoder i run into ambiguous implicits)

**Comment by Tvaroh** [comment]
@johannegger [better](https://failex.blogspot.com.by/2017/04/the-high-cost-of-anyval-subclasses.html?view=sidebar) don't use value classes at all.

**Comment by pjan** [comment]
@Tvaroh not a discussion that we need to have here (I don't want to highjack the conversation), but not everything needs to be heavily speed/memory optimized, and valueclasses can be beneficial as they can eliminate a whole class of developer bugs in environments where these can be particularly costly.

**Comment by Tvaroh** [comment]
@pjan I see your point, but this not only about the performance, it's also about the ability to work with your types programmatically, without leveraging Shapeless-like hacks for that. It's all in the article that I pointed to.

**Comment by mariop-da** [comment]
@Tvaroh nothing in this PR forces the developer to use value classes but for people who may want to use them this can be useful. I don't think this is the right place to have a discussion about a feature of the Scala language.

**Comment by melrief** [comment]
@Tvaroh nothing in this PR forces the developer to use value classes but for people who may want to use them this can be useful. I don't think this is the right place to have a discussion about a feature of the Scala language.

**Comment by fommil** [comment]
@radusw nice, thanks... I'm trying to do something similar for another generic derivation project but I keep getting diverging implicits with this. Where should this go?

**Comment by travisbrown** [comment]
Addressed in #661 and #744 and available in generic-extras in 0.9.0+.
